mindmeld/components/entity_resolver.py

```diff
@@ -562,38 +776,829 @@ def _predict_exact_match(self, entity):
+                    raise ImportError("`torch` library required to quantize models") from None
+
+                torch_qint8 = _get_module_or_attr("torch", "qint8")
+                torch_nn_linear = _get_module_or_attr("torch.nn", "Linear")
+
+                transformer_model = _get_module_or_attr("torch.quantization", "quantize_dynamic")(
+                    transformer_model, {torch_nn_linear}, dtype=torch_qint8
+                ) if transformer_model else None
+                pooling_model = _get_module_or_attr("torch.quantization", "quantize_dynamic")(
+                    pooling_model, {torch_nn_linear}, dtype=torch_qint8
+                ) if pooling_model else None
+                sbert_model = _get_module_or_attr("torch.quantization", "quantize_dynamic")(
+                    sbert_model, {torch_nn_linear}, dtype=torch_qint8
+                ) if sbert_model else None
+
+            setattr(SentenceBertCosSimEntityResolver, "TRANSFORMER_MODEL", transformer_model)
+            setattr(SentenceBertCosSimEntityResolver, "POOLING_MODEL", pooling_model)
+            setattr(SentenceBertCosSimEntityResolver, "SBERT_MODEL", sbert_model)
+
+        self.transformer_model = getattr(SentenceBertCosSimEntityResolver, "TRANSFORMER_MODEL")
+        self.pooling_model = getattr(SentenceBertCosSimEntityResolver, "POOLING_MODEL")
+        self.sbert_model = getattr(SentenceBertCosSimEntityResolver, "SBERT_MODEL")
+
+        msg = "bert embeddings are cached for entity_type: `%s` " \
+              "for quicker entity resolution; consumes some disk space"
+        logger.info(msg, self.type)
+
+    @property
+    def default_er_config(self):
+        defaults = {
+            "model_settings": {
+                "pretrained_name_or_abspath": "distilbert-base-nli-stsb-mean-tokens",
+                "batch_size": 16,
+                "concat_last_n_layers": 4,
+                "normalize_token_embs": True,
+                "bert_output_type": "mean",
+                "augment_lower_case": False,
+                "quantize_model": True,
+                "augment_average_synonyms_embeddings": True
+            }
+        }
+        return defaults
+
+    @property
+    def device(self):
+        return "cuda" if _get_module_or_attr("torch.cuda", "is_available")() else "cpu"
+
+    @property
+    def pretrained_name(self):
+        return os.path.split(self._model_pretrained_name_or_abspath)[-1]
+
+    @staticmethod
+    def cache_path(app_path, er_config, entity_type):
+        """Obtains and return a unique cache path for saving synonyms' embeddings
+
+        Args:
+               er_config: the er_config dictionary of the reolver class
+               entity_type: entity type of the class instance, for unique path identification
+
+        Return:
+            str: path with a .pkl extension to cache embeddings
+        """
+
+        hashid = hashlib.sha1(json.dumps(er_config, sort_keys=True).encode()).hexdigest()
```

Comment:
Do we gaurantee that all the keys are present and filled (ie with even Nones) here? I am concerned about two configs that are equivalent but whose hashes dont match. Example:
```
{'a': '1'}
vs
{'a': '1', 'b': None}
```
Are the two dicts above equivalent for model_configs?

Reply:
That's the reason we have default_er_config, which makes sure all fields are present in er_config. Right now, if a field is inputted as None, we don't overwrite that field with defaults. We just add fields if they are not present in input config.